tests/helpers/cooper_test_utils.py

```diff
@@ -12,6 +12,146 @@
+
+            self.ineq_constraints = cooper.Constraint(
+                constraint_type=cooper.ConstraintType.INEQUALITY,
+                formulation_type=ineq_formulation_type,
+                multiplier=ineq_multiplier_type(
+                    constraint_type=cooper.ConstraintType.INEQUALITY, num_constraints=b.numel(), device=device
+                ),
+                penalty_coefficient=ineq_penalty_coefficient,
+            )
+
+        if has_eq_constraint:
+            eq_penalty_coefficient = None
+            if eq_penalty_coefficient_type is not None:
+                eq_penalty_coefficient = eq_penalty_coefficient_type(init=torch.ones(d.numel(), device=device))
+
+            self.eq_constraints = cooper.Constraint(
+                constraint_type=cooper.ConstraintType.EQUALITY,
+                formulation_type=eq_formulation_type,
+                multiplier=eq_multiplier_type(
+                    constraint_type=cooper.ConstraintType.EQUALITY, num_constraints=d.numel(), device=device
+                ),
+                penalty_coefficient=eq_penalty_coefficient,
+            )
+
+    def _compute_eq_or_ineq_violations(self, x, A_or_C, b_or_d, has_constraint, use_surrogate, multiplier_type):
+        if not has_constraint:
+            return None
+
+        violation = A_or_C @ x - b_or_d
+
+        constraint_features = None
+        if multiplier_type == cooper.multipliers.IndexedMultiplier:
+            constraint_features = torch.randperm(b_or_d.numel(), generator=self.generator, device=self.device)
+            # we assume 80% of the constraints are observed
+            constraint_features = constraint_features[: 4 * b_or_d.numel() // 5]
+            violation = violation[constraint_features]
+
+        strict_violation = None
+        strict_constraint_features = None
+        if use_surrogate:
+            strict_violation = torch.mm(
+                A_or_C + 0.1 * torch.randn(A_or_C.shape, generator=self.generator, device=self.device), x
+            )
+            if multiplier_type == cooper.multipliers.IndexedMultiplier:
+                strict_constraint_features = torch.randperm(
+                    b_or_d.numel(), generator=self.generator, device=self.device
+                )
+                # we assume 80% of the constraints are observed
+                strict_constraint_features = strict_constraint_features[: 4 * b_or_d.numel() // 5]
+                strict_violation = strict_violation[strict_constraint_features]
+
+        return cooper.ConstraintState(
+            violation=violation,
+            constraint_features=constraint_features,
+            strict_violation=strict_violation,
+            strict_constraint_features=strict_constraint_features,
+        )
+
+    def compute_violations(self, x: torch.Tensor) -> cooper.CMPState:
+        """
+        Computes the constraint violations for the given parameters.
+        """
```

Comment:
Add a comment that sampling for strict violation and violation is different

Reply:
TODO?